sopel/modules/currency.py

```diff
@@ -2,109 +2,201 @@
-    ''', re.VERBOSE)
+FIAT_URL = 'https://api.exchangeratesapi.io/latest?base=EUR'
+FIXER_URL = 'http://data.fixer.io/api/latest?base=EUR&access_key={}'
+CRYPTO_URL = 'https://apiv2.bitcoinaverage.com/indices/global/ticker/short?crypto=BTC'
+EXCHANGE_REGEX = re.compile(r'''
+    ^(\d+(?:\.\d+)?)                                            # Decimal number
+    \s*([a-zA-Z]{3})                                            # 3-letter currency code
+    \s+(?:in|as|of|to)\s+                                       # preposition
+    (([a-zA-Z]{3}$)|([a-zA-Z]{3})\s)+$                          # one or more 3-letter currency code
+''', re.VERBOSE)
+LOGGER = get_logger(__name__)
+UNSUPPORTED_CURRENCY = "Sorry, {} isn't currently supported."
+UNRECOGNIZED_INPUT = "Sorry, I didn't understand the input."
 
+rates_fiat_json = {}
+rates_btc_json = {}
+rates_updated = 0.0
 
-def get_rate(code):
-    code = code.upper()
-    if code == 'CAD':
-        return 1, 'Canadian Dollar'
-    elif code == 'BTC':
-        btc_rate = get('https://apiv2.bitcoinaverage.com/indices/global/ticker/BTCCAD')
-        rates = btc_rate.json()
-        return 1 / rates['averages']['day'], 'Bitcoin—24hr average'
 
-    data = get(base_url.format(code))
-    name = data.json()['seriesDetail']['FX{}CAD'.format(code)]['description']
-    name = name.split(" to Canadian")[0]
-    json = data.json()['observations']
-    for element in reversed(json):
-        if 'v' in element['FX{}CAD'.format(code)]:
-            return 1 / float(element['FX{}CAD'.format(code)]['v']), name
+class CurrencySection(StaticSection):
+    fixer_io_key = ValidatedAttribute('fixer_io_key', default=None)
+    auto_convert = ValidatedAttribute('auto_convert', parse=bool, default=False)
 
 
-@commands('cur', 'currency', 'exchange')
-@example('.cur 20 EUR in USD')
-def exchange(bot, trigger):
+def configure(config):
+    config.define_section('currency', CurrencySection, validate=False)
+    config.currency.configure_setting('fixer_io_key', 'Optional API key for Fixer.io (leave blank to use exchangeratesapi.io):')
+    config.currency.configure_setting('auto_convert', 'Automatically respond to currency conversion lines?')
+
+
+def setup(bot):
+    bot.config.define_section('currency', CurrencySection)
+
+
+def update_rates(bot):
+    global rates_fiat_json, rates_btc_json, rates_updated
+
+    # If we have data that are less than 24h old, return
+    if time.time() - rates_updated < 24 * 60 * 60:
+        return
+
+    # Update crypto rates
+    response = requests.get(CRYPTO_URL)
+    response.raise_for_status()
+    rates_btc_json = response.json()
+
+    # Update fiat rates
+    if bot.config.currency.fixer_io_key is not None:
+        response = requests.get(FIXER_URL.format(bot.config.currency.fixer_io_key))
+        if not response.json()['success']:
+            LOGGER.error(str(response.json()['error']))
+            return bot.reply('Sorry, something went wrong')
```

Comment:
Have this raise an exception so you can differentiate this `return` (en error) from the `return` when data less than 24h old already exist (a normal continuation).